RandLAPACK/comps/rl_preconditioners.hh

```diff
@@ -0,0 +1,208 @@
+#ifndef randlapack_comps_preconditioners_h
+#define randlapack_comps_preconditioners_h
+
+#include "rl_orth.hh"
+#include "rl_rf.hh"
+#include "rl_blaspp.hh"
+#include "rl_lapackpp.hh"
+#include "rl_util.hh"
+
+#include <RandBLAS.hh>
+#include <math.h>
+#include <cstdint>
+#include <limits>
+#include <vector>
+
+#include <iostream>
+#include <cstdint>
+
+namespace RandLAPACK {
+
+/**
+ * Use sketching to produce data for a right-preconditioner for tall
+ * matrices of the form A_aug = [A; sqrt(mu) * I]. This is useful for ...
+ * 
+ *      Overdetermined and underdetermined least squares
+ *      ------------------------------------------------
+ *      It makes A_pre = A_aug*M well-conditioned, so that
+ *      min{||A_pre * z - b ||} and min{ ||y|| s.t. A_pre' y = c }
+ *      can easily be solved by unpreconditioned iterative methods.
+ * 
+ *      Solving linear systems with G = A'A + mu*I
+ *      ------------------------------------------
+ *      It makes M'GM well-conditioned, so that M'GM z = M'h
+ *      can easily be solved by CG without preconditioning.
+ * 
+ * The input matrix A must be contiguous; no "LDA" is allowed.
+ * 
+ * On exit, the preconditioner with mu=0 would be defined by
+ * taking the columns of V_sk and scaling them by 1/sigma_sk.
+ * We do not perform this scaling ourselves, since other values
+ * of mu might be intended.
+ * 
+ * @param[in] layout_A
+ *      Either blas::Layout::RowMajor or blas::Layout::ColMajor.
+ * @param[in] m
+ *      The number of rows in A
+ * @param[in] n
+ *      The number of columns in A (and rows in M)
+ * @param[in] d
+ *      The number of rows in the sketched data matrix; must be >= n.
+ * @param[in] k
+ *      The number of nonzeros per column in the sketching operator
+ *      A common value is k=8 when n is on the order of a couple thousand.
+ * @param[in] A
+ *      A buffer for an m \times n matrix A. Interpreted in 
+ *      "layout_A" order.
+ * @param[out] V_sk
+ *      We require that V_sk.size() >= d*n. Letting "r" denote the return
+ *      value of this function, on exit the orthogonal factor of the preconditioner
```

Comment:
This function returns void. Unclear what r represents.

Reply:
Ah crap. Will fix.

Reply:
Fixed. I've updated the doc to not mention this rank parameter and to have this function return an RNGState.